New Email Magic Links Overview page (#3453)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/email-magic-links-overview/main/index.md

## Integrate Magic Links

Integrating Email Magic Links into your application is a two-step process.

1. Create an endpoint for your magic link. This endpoint checks the validity of the OTP and sends it to Identity Engine for validation.
1. Set that endpoint as the callback URL for magic links for your application with the Admin Console for your org.

### Create an endpoint for your magic links

Any magic link URL pointing to your application contains two query parameters:

* The one-time password that validates the user (`otp`)
* A state token that uniquely identifies the current authentication process and its state (`state`)

For example, `http://${yourOktaDomain}?otp=726009&state=1b31fa98b34c45d9a`.

You must create an endpoint for your application that:

1. Retrieves `otp` and `state` values from the query parameters
2. Matches the state token with the current `state` in your user's browser session
3. Makes any other checks you deem necessary to ensure that the user is working in the same browser on the same device
4. Requests that the user enter the `otp` value manually if steps 2 or 3 fail

> 5. Sends `otp` and `state` to Identity Engine for validation if steps 2 and 3 pass

Reviewer: the `state` in step 5 is the Idx API `state handle`. I'm not aware of this concept or the Idx API being documented yet. Step 5 seems confusing as it reads as the `state` is the same as in step 1 and it is not. The SDKs will use the `state` from step 1 to find the idx state handle to be used on step 5.

---

> title: Email Magic Links Overview

Author: done

---

<ApiLifecycle access="ie" /><br />

> Enable the user to enter a one-time password in a single step by clicking an embedded email link.

Author: fixed

> ## Introduction

Author: @robertnicolo-okta  We don't have nutrition facts in overview docs, do we?

> ## Introduction

Reviewer 2: Since this was a beefy overview doc, I thought nutrition facts would be helpful here.  I would confirm with Chris on this.

> When using email to prove their identity, users can copy a one-time password (OTP) from an email into the application they want to use. Email Magic Links (EML) provides a second option, allowing users to click a hyperlink in the email rather than use the OTP - a quicker, more user-convenient, and still secure experience. However, applications based on the Embedded Sign-In Widget or an Embedded SDK must be adapted to make use of EML.

Author: fixed

> When using email to prove their identity, users can copy a one-time password (OTP) from an email into the application they want to use. Email Magic Links (EML) provides a second option, allowing users to click a hyperlink in the email rather than use the OTP - a quicker, more user-convenient, and still secure experience. However, applications based on the Embedded Sign-In Widget or an Embedded SDK must be adapted to make use of EML.

Author: No it's fine. I didn't know this. Have changed to your suggestion.

<div class="half">

![A simple diagram showing a cursor clicking a magic link in an email and beign signed in successfully](/img/advanced-use-cases/email-magic-link-overview-simple-summary.png)

</div>

> In this overview, you will learn:

Author: changed to Learning Outcomes

* How using EML works compared to using OTP

> * How to integrate EML into applications that use the Embedded Sign-In Widget or an Embedded SDK.

Author: updated

> > Note: Email Magic Links supports [Embedded Sign-In Widget](https://github.com/okta/okta-signin-widget) v6.3.5 or later, and the following Embedded SDKs:

Author: Confirmed with Chris, thsi is fine.

>
> * [Auth.js](https://github.com/okta/okta-auth-js) v6.0.0
> * [Java IDX SDK](https://github.com/okta/okta-idx-java) v2.0.1
> * [.NET IDX SDK](https://github.com/okta/okta-idx-dotnet) v2.1.0
>

> > We recommend using the latest version of the Sign-In Widget or selected SDK.

Author: fixed

## Email Magic Links vs. one-time passwords

The Okta email authenticator provides a simple way for users to authenticate themselves or recover their accounts. It supports two methods for users to validate their credentials as part of those processes. In both methods, an email is sent to their primary email address and either:

1. The user copies an OTP from the email to their application and submits it to Identity Engine for authentication.
1. The user clicks a link in the email that submits the OTP to Identity Engine for authentication on their behalf.

> If the OTP is valid and submitted within a set time, either the user is authenticated, or they continue their authentication process. The following diagram illustrates the user's experience with the two options.

Author: nice. updated

> If the OTP is valid and submitted within a set time, either the user is authenticated, or they continue their authentication process. The following diagram illustrates the user's experience with the two options.

Author: fixed

<div class="full">

> ![Flow diagrams showing the difference in user experience between using OTP and using magic links for authentication](/img/advanced-use-cases/email-magic-links-overview-user-flow-comparison.png)

Author: No idea. Just copied the original diagram from the SSPR article. Have removed "default behavior".

</div>

> The OTP experience involves an extra step for the user - copying the OTP into the application - but it does ensure they return to the same browser to continue their authentication process. The magic link experience, on the other hand, has fewer steps. Clicking the magic link opens a new tab in the user's browser and their authentication process carries on there.

Author: fixed

> The OTP experience involves an extra step for the user - copying the OTP into the application - but it does ensure they return to the same browser to continue their authentication process. The magic link experience, on the other hand, has fewer steps. Clicking the magic link opens a new tab in the user's browser and their authentication process carries on there.

Author: updated

### Same device, same browser only

Magic links only work when there is complete assurance that the person who started the request is the same one who clicked the magic link. For this reason, a user who starts signing in to your application in a web browser must be in the same browser on the same device when they click the magic link. If their browser or device is different, the magic link is disabled. They must either return to the original browser to use the magic link or use the OTP instead. The following flowchart illustrates this logic.

<div class="full">

> ![Flow diagram showing when and how the magic link flow falls back to the OTP flow](/img/advanced-use-cases/email-magic-links-overview-fallback-to-otp.png)

Author: Fixed 1. Discussed 2 adn 3 with you. Stet.

</div>

This same flow also ensures that users don't sign in to applications from non-managed devices. For example, should a user start the sign-in process on a company laptop, and then click the magic link from their personal device, they must return to the company laptop to complete the process.

> ### Pros and Cons

Author: Pros and Cons of using EML....

> ### Pros and Cons

Reviewer 3: should we be more specific here and say that? Pros and Cons of EML?

> ### Pros and Cons

Author: stet

> When evaluating whether to support magic links in addition to OTPs in the email authenticator, you should consider the following:

Author: nice. changed

* Users get a better sign-in experience from magic links than OTP if they sign in from the same browser on the same device.

> * If users sign in from a different browser or a different device, the process falls back to using an OTP.

Author: fixed

> * An application can only be accessed by clicking a magic link on a device that already has access to that application.

Reviewer 3: Avoid using sign-in as a noun:

A user can only successfully authenticate from a magic link if the same browser and device are used to initiate the sign-in flow.

> * An application can only be accessed by clicking a magic link on a device that already has access to that application.

Author: fixed

> * An application can only be accessed by clicking a magic link on a device that already has access to that application.

Author: Rob had the same change :-)

> * Okta provides developer support for both OTP and Magic Links in their Embedded Sign-In Widget and Identity Engine SDKs.

Author: fixed

> ## Integrating Magic Links

Author: Noted

Integrating Email Magic Links into your application is a two-step process.

> 1. Create an endpoint for your magic link. This checks the validity of the OTP and sends it to the Identity Engine for validation.

Author: fixed

> 1. Create an endpoint for your magic link. This checks the validity of the OTP and sends it to the Identity Engine for validation.

Author: Yup

1. Set that endpoint as the callback URL for magic links for your application.

### Create an endpoint for your magic links

Any magic link URL pointing to your application contains two query parameters:

* The one-time password that validates the user (`otp`)

> * A state token that uniquely identifies the current authentication process and the state it is in (`state`).

Author: fixed

> * A state token that uniquely identifies the current authentication process and the state it is in (`state`).

Author: a rare removal of punctuation :-)

For example, `http://${yourOktaDomain}?otp=726009&state=1b31fa98b34c45d9a`.

You must create an endpoint for your application that:

1. Retrieves `otp` and `state` values from the query parameters
2. Matches the state token with the current `state` in your user's browser session

> 3. Makes any other checks you deem necessary to ensure the user is working in the same browser on the same device

Author: fixed

4. Requests the user enter the `otp` value manually if steps 2 or 3 fail
5. Sends `otp` and `state` to Identity Engine for validation if steps 2 and 3 pass
6. Redirects the user to a page that continues their authentication process

> If your application uses the Embedded Sign-In Widget to authenticate users, pass `otp` and `state` to it as you instantiate it on the page. The Sign-In Widget handles steps 2 to 6.

Author: It doesn't in .NET either. I've reworded to that effect and will link to your Sign In With Email Only Widget guide when that is up for full details. Is that OK?

> If your application uses the Embedded Sign-In Widget to authenticate users, pass `otp` and `state` to it as you instantiate it on the page. The Sign-In Widget handles steps 2 to 6.

Reviewer 2: yes

```javascript
var searchParams = new URL(window.location.href).searchParams;
var signIn = new OktaSignIn({
   baseUrl: '${yourOktaDomain}',
   baseUrl: '${yourOktaDomain}',
   clientId: '${yourClientId}',
   redirectUri: '${yourSignInRedirectUrl}',
   useInteractionCodeFlow: true,
   otp: searchParams.get('otp'),
   state: searchParams.get('state')
});
```

If your application uses a supported Embedded SDK to authenticate users, the [Okta email integration](https://developer.okta.com/docs/guides/authenticators-okta-email/aspnet/main/) and [custom password recovery](https://developer.okta.com/docs/guides/oie-embedded-sdk-use-case-custom-pwd-recovery-mfa/aspnet/main/) guides cover how to implement these steps.

### Point your app's magic links to that endpoint

> There are two ways to set your application's magic links to the endpoint you have created:

Author: fixed

* You can set your application's Email Verification Experience (EVE) value to the endpoint's URL.
* You can customize each of the email templates used by the email authenticator to include the endpoint's URL.

> Changing the EVE provides a quick way to enable and test magic links. Customizing the email templates requires more work but results in fewer roundtrips to the Identity Engine and a faster user experience.

Reviewer 3: @robertnicolo-okta no articles are used in front of Identity Engine. That has changed, sadly. I did an across the board fix of that and updated the style guide back when that change was made.

> Changing the EVE provides a quick way to enable and test magic links. Customizing the email templates requires more work but results in fewer roundtrips to the Identity Engine and a faster user experience.

Author: I deleted "the". I suppose "the Identity Engine backend" or "servers" woudl work to, but IE is all backend so....  I just added in the the without thinking too much.

> #### Using the Email Verification Experience

Author: fixed

To enable magic links in your application using the email verification experience (EVE) setting:

1. Open the **Admin Console** for your Okta org.
2. Choose **Applications > Applications** to show the app integrations that you have already created.
3. Click the application that you previously created.
4. In the **General Settings** section on the **General** tab, click **Edit**.
5. Under **EMAIL VERIFICATION EXPERIENCE** enter the URL for the endpoint you created. For example, `https://example.com/magiclink/callback`.
6. Click **Save** to save your changes.

When required, Identity Engine sends an email similar to this one containing an OTP and a link:

<div class="half border">

> ![A sample email sent by the Identity Engine with OTP and magic link highlighted](/img/advanced-use-cases/email-magic-links-overview-sample-email.png)

Author: I've deleted the border style. Whatever we end up doing with the images I don't imagine it will involve putting a border all the way aroudn the edge of this image.

</div>

> 

Author: done

> When a user clicks the magic link from the email authenticator, a request is sent to the Identity Engine to validate the user. It has a token parameter and looks something like this:

Author: done.

`https://${yourOktaDomain}/email/verify/0oa1esny76wdQqjsg697?token=ftlNKCIYA_4WvR6BPbYhyKisVYIF-QLp2t`

> Authentication continues at an Okta-hosted page if the EVE setting doesn't have a value. If the EVE setting has a value, Identity Engine builds a new URL based on the EVE setting and the current `otp` and `state` values. Then it redirects the user's browser to that URL. For example:

Author: done

`https://example.com/magiclink/callback?state=1ed1d4de1a50b8f30d87e720bcc1b6a3&otp=439246`

After validating the `otp` and `state` parameters, the authentication process continues for the user.

<div class="full">

![Flow diagram showing the interaction between application and Identity Engine when using the Verification Experience setting](/img/advanced-use-cases/email-magic-links-overview-experience-flow.png)

</div>

The EVE setting offers a quick way to redirect magic links to your application, but it does require a round trip to the Identity Engine to retrieve the correct URL for the endpoint you have created. Unless you are willing to customize the email templates the email authenticator uses, the user receives only Okta-branded emails.

If your user base is multilingual, consider that the Okta-branded email templates support many languages by default. After you customize an email template, you must create a copy of that template in each language you support.

> #### Using custom email templates

Author: @robertnicolo-okta I'm not comfortable with adding this. You're not suggesting adding example to the EVE heading, so why this one? We don't use "example" in any of our other guides.

> #### Using custom email templates

Author: fixed

> #### Using custom email templates

Reviewer 2: @danielmaharry-okta In this section, you are discussing the Email Challenge template only. My aim with "example" was  just to make it explicit that not all templates use the same variables (ie. emailAuthenticationLink), HTML ID, etc.

Creating custom authentication email templates allows you to send a fully-branded email to your users that skips the round-trip to Identity Engine and redirects users straight to your endpoint.

All Okta email templates are written using [Velocity Templating Language (VTL)](https://help.okta.com/en-us/Content/Topics/Settings/velocity-variables.htm) and use predefined variables to insert relevant values into that email. For example, Okta defines three VTL variables specific to the **Email Challenge** template:

| Variable                     | Contains                                                       |
|------------------------------|----------------------------------------------------------------|

> | `${verificationToken}`       | The one-time password Okta generated for the user              |

Reviewer 3: ```suggestion
| `${verificationToken}`       | The one-time password that Okta generated for the user              |
```

Author: Changed to "that Identity Engine"

| `${request.relayState}`      | The state token for the current authentication process         |
| `${emailAuthenticationLink}` | The Okta-hosted URL that continues the password recovery flow  |

> By default, the magic link in the template is set to `${emailAuthenticationLink}`. If you're using the Email Verification Experience setting for your redirects, you don't need to change this. The magic link will work as explained in the [Using the Email Verification Experience](#using-the-email-verification-experience) section.

Reviewer 3: ```suggestion
By default, the magic link in the template is set to `${emailAuthenticationLink}`. If you're using the Email Verification Experience setting for your redirects, you don't need to change this. The magic link works as explained in the [Use the Email Verification Experience](#use-the-email-verification-experience) section.
```

Author: fixed

To point the magic link directly to the endpoint in your application, you must replace `${emailAuthenticationLink}` with your custom URL that includes `otp` and `state` as query parameters:

> 1. In the **Admin Console**, go to **Customizations > Emails**.

Reviewer 3: ```suggestion
1. In the **Admin Console**, go to **Customizations** > **Emails**.
```

Author: fixed

2. On the **Emails** page, find the **Other** category on the template menu.
3. Under **Other**, click **Email Challenge**.
4. On the **Email Challenge** email template page, click **Edit**.
5. Under **Default Email**, click **Edit**.
6. In the Message field, locate the magic link in the field's HTML. The link is located in the href attribute of an `<a>` tag with the `id` of `email-authentication-button`. It looks like this:

   ```html
   <a id="email-authentication-button"
   href="${emailAuthenticationLink}"
   style="text-decoration: none;">
      <span style="padding: 9px ...;">
         Sign In
      </span>
   </a>
   ```

7. Replace the `${emailAuthenticationLink}` variable with the URL for your endpoint. You must append the `${verificationToken}` and `${request.relayState}` variables as query parameter values. For example:

   ```html
   <a id="email-authentication-button"
   href="https://example.com/magiclink/callback?otp=${verificationToken}&state=${request.relayState}"
   style="text-decoration: none;">
      <span style="padding: 9px ...;">
         Sign In
      </span>
   </a>
   ```

8. Click **Save** and close the dialog.

> Four email templates can be customized in this way. These are listed in the table below.

Reviewer 3: ```suggestion
You can customize four email templates this way. These are listed in the table below.
```

> Four email templates can be customized in this way. These are listed in the table below.

Author: done

> * All support redirects via EVE and direct redirects.

Reviewer 3: ```suggestion
* All support redirects through EVE and direct redirects.
```

Author: fixed

* All are compatible with applications based on either the Embedded Sign-In Widget or an Embedded SDK.

> * To add `otp` and `state` variables to the magic link, find the template you are editing in the table, and:

Reviewer 3: ```suggestion
* To add `otp` and `state` variables to the magic link, find the template that you are editing in the table, and:
```

Author: added

  * Replace the VTL variable in the **magic link variable name** column with your custom endpoint URL.
  * Set your `otp` query string variable to the VTL variable in the **OTP variable name** column for the template.

>   * Set your `state` query string variable to `${request.relayState}` for all templates.

Author: Doesn't seem to be anythign to fix here.

> | Template Name  | Use Case | Magic link variable name | OTP variable name |

Reviewer 3: ```suggestion
| Template name  | Use case | Magic link variable name | OTP variable name |
```

Author: changed

|----------------|----------|--------------------------|-------------------|
| Email challenge | Sign in with email - challenge | `${emailAuthenticationLink}`    | `${verificationToken}` |
| Forgot Password | Self-service password recovery | `${resetPasswordUrl}`           | `${oneTimePassword}`   |
| Registration - Activation | Self-service registration | `${registrationActivationLink}` | `${oneTimePassword}`   |
| Email factor verification | Sign in with email - enrollment | `${verificationLink}` | `${verificationToken}` |

When a user clicks the magic link based on a customized email template, their browser is redirected straight to your application's endpoint:

<div class="full">

![Flow diagram showing the interaction between application and Identity Engine when using custom email templates](/img/advanced-use-cases/email-magic-links-overview-custom-template-flow.png)

</div>

### Comparing the options

> |   | EVE + Default email templates | EVE + Custom email templates | Custom Email Templates |

Reviewer 3: ```suggestion
|   | EVE + Default email templates | EVE + Custom email templates | Custom email templates |
```

Author: changed

|---|---|----|----|

> | Set-up + Maintenance   | Requires one change in the admin console | Requires editing every email template that uses magic links to point to your new endpoint. | Requires editing every email template that uses magic links to point to your new endpoint. |

Reviewer 3: ```suggestion
| Set-up + Maintenance   | Requires one change in the Admin Console | Requires editing every email template that uses magic links to point to your new endpoint. | Requires editing every email template that uses magic links to point to your new endpoint. |
```

Author: fixed

| Speed of redirection   | Magic link sends a request to Identity Engine for the callback URL. The user is then redirected to the callback URL. | Magic link sends a request to Identity Engine for the callback URL. The user is then redirected to the callback URL. | Magic link sends browser directly to the callback URL |
| Branding | Default Okta-branded email templates | You can add your brand to a custom email template  | You can add your brand to a custom email template |
| Multi-language support | Default email templates support multiple languages | You must create and maintain a version of each custom template in each required language. | You must create and maintain a version of each custom template in each required language. |

## See also

> * [Okta Email (magic link/OTP) integration guide](https://developer.okta.com/docs/guides/authenticators-okta-email/-/main/)

Reviewer 2: Two changes:

1. Remove the domain names and start with the Url with "docs"
2. I understand why the tilda is used but clicking on the links bounces you out of the left nav bar which is less than optimal. Possibility default to a language?

Author: I've swapped out all the guide links into separate stack snippets so they all have their won correct links.

---
title: Email Magic Links Overview
---

<ApiLifecycle access="ie" /><br />

Enable applications based on the Embedded Widget or an Embedded SDK to authenticate users with a single URL click from an email.

## Introduction

When using email to prove their identity, users can copy a one-time password (OTP) from an email into the application they want to use. Email Magic Links (EML) provides a second option, allowing users to click a hyperlink in the email rather than use the OTP - a quicker, more user-convenient, and still secure experience. However, applications based on the Embedded Widget or an Embedded SDK must be adapted to make use of EML.

<div class="half">

![A simple diagram showing a cursor clicking a magic link in an email and beign signed in successfully](/img/advanced-use-cases/email-magic-link-overview-simple-summary.png)

</div>

In this overview, you will learn:

* How using EML works compared to using OTP
* How to integrate EML into applications that use the Embedded Widget or an Embedded SDK.

> > NOTE: Email Magic Links was first supported in [Embedded Sign-In Widget](https://github.com/okta/okta-signin-widget) v6.3.5, and the following Embedded SDKs: [Auth.js](https://github.com/okta/okta-auth-js) v6.0.0, [Java IDX SDK](https://github.com/okta/okta-idx-java) v2.0.1, and [.NET IDX SDK](https://github.com/okta/okta-idx-dotnet) v2.1.0. These versions should be used as a minimum to support EML. We recommend using the latest versions.

Author: fixed

</div>

The OTP experience involves an extra step for the user - copying the OTP into the application - but it does ensure they return to the same browser to continue their authentication process. The magic link experience, on the other hand, has fewer steps. Clicking the magic link opens a new tab in the user's browser and their authentication process carries on there.

### Same device, same browser only

Magic links only work when there is complete assurance that the person who started the request is the same one who clicked the magic link. For this reason, a user who starts signing in to your application in a web browser must be in the same browser on the same device when they click the magic link. If their browser or device is different, the magic link is disabled. They must either return to the original browser to use the magic link or use the OTP instead. The following flowchart illustrates this logic.

<div class="full">

![Flow diagram showing when and how the magic link flow falls back to the OTP flow](/img/advanced-use-cases/email-magic-links-overview-fallback-to-otp.png)

</div>

> This same flow also ensures that users do not sign in to applications from non-managed devices. For example, should a user start the sign-in process on a company laptop, and then click the magic link from their personal mobile, they will be directed to return to the company laptop and try again.

Author: fixed

### Pros and Cons

When evaluating whether to support magic links in addition to OTPs in the email authenticator, you should consider the following:

* Users get a better sign-in experience from magic links than OTP if they sign in from the same browser on the same device.
* If users sign in from a different browser or a different device, the process may fall back to using an OTP.

> * An application cannot be accessed by clicking a magic link on a device that does not already have access to that application.

Author: fixed

* Okta provides developer support for both OTP and Magic Links in their Embedded Widget and Identity Engine SDKs.

## Integrating Magic Links

Integrating Email Magic Links into your application is a two-step process.

1. Create an endpoint for your magic link. This checks the validity of the OTP and sends it to the Identity Engine for validation.
1. Set that endpoint as the callback URL for magic links for your application.

### Create an endpoint for your magic links

> Any magic link URL pointing to your application will contain two query parameters

Author: fixed

> * the one-time password that will validate the user (`otp`)

Author: fixed

> * a state token that uniquely identifies the current authentication process and the state it is in (`state`). This is the current SAML [relaystate](https://developer.okta.com/docs/concepts/saml/#understanding-sp-initiated-sign-in-flow) value.

Reviewer: `request.relayState` not available for SAML flows. The `state` in this case is the OIDC/OAuth2 state parameter defined by the client for the current Authorization request. It is covered in the doc Ian pointed at and also in https://developer.okta.com/docs/guides/implement-grant-type/authcodepkce/main/#next-steps and probably other places.

> * a state token that uniquely identifies the current authentication process and the state it is in (`state`). This is the current SAML [relaystate](https://developer.okta.com/docs/concepts/saml/#understanding-sp-initiated-sign-in-flow) value.

Reviewer 4: I defer to @zeusandrade-okta on this. I think if it we just cut it off after "...the state it is in (`state`)." That would be enough for this doc.

For example, `http://${yourOktaDomain}?otp=726009&state=1b31fa98b34c45d9a`.

> You must create an endpoint for your application that

Author: fixed

1. Retrieves `otp` and `state` values from the query parameters
2. Matches the state token with the current `state` in your user's browser session
3. Makes any other checks you deem necessary to ensure the user is working in the same browser on the same device
4. Requests the user enter the `otp` manually if steps 2 or 3 fail their checks
5. Sends `otp` and `state` to Identity Engine to be validated if steps 2 and 3 pass their checks
6. Redirects the user to a page that continues their authentication process

> If your application uses the Embedded Sign-In Widget to authenticate users, pass `otp` and `state` to it as you instantiate it on the page. The Widget itself does steps 2 to 6 for you.

Author: fixed

* [.NET IDX SDK](https://github.com/okta/okta-idx-dotnet) v2.1.0

These versions should be used as a minimum to support EML, but you should prefer to use the latest versions.

### Point your app's magic links to that endpoint

There are two ways to set your application's magic links to the endpoint you have created:

* You can set your application's Email Verification Experience (EVE) value to the endpoint's URL.
* You can customize each of the email templates used by the email authenticator to include the endpoint's URL.

Changing the EVE provides a quick way to enable and test magic links. Customizing the email templates requires more work but results in fewer roundtrips to the Identity Engine and a faster user experience.

#### Using the Email Verification Experience

To enable magic links in your application using the email verification experience (EVE) setting:

1. Open the **Admin Console** for your Okta org.
2. Choose **Applications > Applications** to show the app integrations that you have already created.
3. Click the application that you previously created.
4. In the **General Settings** section on the **General** tab, click **Edit**.

> 5. Under **EMAIL VERIFICATION EXPERIENCE** enter the URL for the endpoint you created. For example, `http://example.com/magiclink/callback`.

Author: I'm not 100% this is true, but better to be on the safe side. I think one of the sample apps doesn't require it but that might be my poor recall.

6. Click **Save** to save your changes.

> When required, Identity Engine sends an email similar to this one containing an OTP and a link.

Author: fixed.

> <div class="half border">

Author: @IanHakes-okta  The half style just got fixed to include a 24px border-bottom so this is fixed elsewhere.

![A sample email sent by the Identity Engine with OTP and magic link highlighted](/img/advanced-use-cases/email-magic-links-overview-sample-email.png)

</div>

> When a user clicks the magic link from the email authenticator, a request is sent to the Identity Engine to validate the user. It has a token parameter and will look something like this:

Author: fixed

`https://${yourOktaDomain}/email/verify/0oa1esny76wdQqjsg697?token=ftlNKCIYA_4WvR6BPbYhyKisVYIF-QLp2t`

> Authentication continues at an Okta-hosted page if the EVE setting does not have a value. If the EVE setting has a value, Identity Engine builds a new URL based on the EVE setting and current `otp` and `state` values. Then it redirects the user's browser to that URL. For example

Author: fixed

`https://example.com/magiclink/callback?state=1ed1d4de1a50b8f30d87e720bcc1b6a3&otp=439246`

> Their authentication process continues there.

Author: fixed

<div class="full">

![Flow diagram showing the interaction between application and Identity Engine when using the Verification Experience setting](/img/advanced-use-cases/email-magic-links-overview-experience-flow.png)

</div>

> The EVE setting offers a quick way to redirect magic links to your application, but it does require a round trip to the Identity Engine to retrieve the correct URL for the endpoint you have created. Unless you are willing to customize the email templates the email authenticator uses, the user receives only Okta-branded emails. On the other hand, if your user base is multilingual, the default Okta templates support many languages by default. Once you customize an email template, you must create a copy for each language you want to have available.

Author: Thansk for the rewording. Definitely clearer :-)

#### Using custom email templates

Creating custom authentication email templates allows you to send a fully-branded email to your users that skips the round-trip to Identity Engine and redirects users straight to your endpoint.

All Okta email templates are written using [Velocity Templating Language (VTL)](https://help.okta.com/en-us/Content/Topics/Settings/velocity-variables.htm) and use predefined variables to insert relevant values into that email. For example, Okta defines three VTL variables specific to the **Email Challenge** template:

| Variable                     | Contains                                                       |
|------------------------------|----------------------------------------------------------------|
| `${verificationToken}`       | The one-time password Okta generated for the user              |

> | `${request.relayState}`      | The current SAML relaystate value                              |

Reviewer 4: Is this variable not just for SAML - this is an OAuth parameter IIRC and OIDC protocol doesn't support it directly. Pinging @zeusandrade-okta or @moushmibanerjee-okta to review and clarify this point.

Author: I have changed this to " The state token for the current authentication process". I hope that is OK?

| `${emailAuthenticationLink}` | The Okta-hosted URL that continues the password recovery flow  |

> By default, the magic link in the template is set to `${emailAuthenticationLink}`. If you're using the Email Verification Experience setting for your redirects, you don't need to change this. The magic link will work as explained in the Using the Email Verification Experience section.

Author: done

> To point the magic link directly to the endpoint in your application, you must replace `${emailAuthenticationLink}` with your custom URL that includes `otp` and `state` as query parameters.

Author: fixed

2. On the **Emails** page, find the **Other** category on the template menu.
3. Under **Other**, click **Email Challenge**.
4. On the **Email Challenge** email template page, click **Edit**.
5. Under **Default Email**, click **Edit**.
6. In the Message field, locate the magic link in the field's HTML. The link is located in the href attribute of an `<a>` tag with the `id` of `email-authentication-button`. It looks like this:

   ```html
   <a id="email-authentication-button"
   href="${emailAuthenticationLink}"
   style="text-decoration: none;">
      <span style="padding: 9px ...;">
         Sign In
      </span>
   </a>
   ```

7. Replace the `${emailAuthenticationLink}` variable with the URL for your endpoint. You must append the `${verificationToken}` and `${request.relayState}` variables as query parameter values. For example:

   ```html
   <a id="email-authentication-button"
   href="https://example.com/magiclink/callback?otp=${verificationToken}&state=${request.relayState}"
   style="text-decoration: none;">
      <span style="padding: 9px ...;">
         Sign In
      </span>
   </a>
   ```

8. Click **Save** and close the dialog.

> Nine email templates can be customized in this way. These are listed in the table below.

Reviewer 4: Change to four email templates

Author: done

  * Set your `state` query string variable to `${request.relayState}` for all templates.

| Template Name                                  | Use Case                                                     | Magic link variable name      | OTP variable name    |
|------------------------------------------------|--------------------------------------------------------------|-------------------------------|----------------------|
| Email challenge                                | Sign in with email - challenge                               | `${emailAuthenticationLink}`    | `${verificationToken}` |
| Forgot Password                                | Self-service password recovery                               | `${resetPasswordUrl}`           | `${oneTimePassword}`   |
| Registration - Activation                      | Self-service registration                                    | `${registrationActivationLink}` | `${oneTimePassword}`   |
| Email factor verification                      | Sign in with email - enrollment                              | `${verificationLink}`           | `${verificationToken}` |

> | Self-service unlock account\*                   | Self-service account unlock                                  | `${unlockAccountLink}`          | `${oneTimePassword}`   |

Author: I've removed as requested

| Active Directory Forgot Password\*              | Self-service password recovery for Active Directory accounts | `${resetPasswordUrl}`           | `${oneTimePassword}`   |
| Active Directory Self-Service Unlock Account\*  | Self service account unlock for Active Directory accounts    | `${unlockAccountLink}`          | `${oneTimePassword}`   |
| LDAP Forgot Password\*                          | Self-service password recovery for LDAP accounts             | `${resetPasswordUrl}`           | `${oneTimePassword}`   |
| LDAP Self-Service Unlock Account\*              | Self-service account unlock for LDAP accounts                | `${unlockAccountLink}`          | `${oneTimePassword}`   |

> \* Available from September 2022.

Author: done

> When a user clicks the magic link based on a customized email template, their browser is redirected straight to your application's endpoint.

Author: ok

<div class="full">

![Flow diagram showing the interaction between application and Identity Engine when using custom email templates](/img/advanced-use-cases/email-magic-links-overview-custom-template-flow.png)

</div>

### Comparing the options

|   | EVE + Default email templates | EVE + Custom email templates | Custom Email Templates |
|---|---|----|----|

> | Set-up + Maintenance   | Requires one change in the admin console | Requires editing every email template you want to contain magic links pointing to your new endpoint. | Requires editing every email template to contain magic links pointing to your new endpoint. |

Author: fixed

| Speed of redirection   | Magic link sends a request to Identity Engine for the callback URL. The user is then redirected to the callback URL. | Magic link sends a request to Identity Engine for the callback URL. The user is then redirected to the callback URL. | Magic link sends browser directly to the callback URL |

> | Branding | Default email templates are Okta branded | You can add your brand to a custom email template however you like  | You can add your brand to a custom email template however you like |

Author: done
